[CBRD-23903] Set tuple_count as INT64 type in list file

hgryoo opened the pull request:
http://jira.cubrid.org/browse/CBRD-23903

At the previous PR(#2666), The return type of count queries is changed to BIGINT type. However, counting row count should be performed in real.
```
All kinds of queries to count # of rows are gathered into list_file.tuple_count

                      Temp file, including result (list_file)
                                       ^
    -----------------------------------|----------------------------------------------
    |                                  |                   |                         |
B-tree(index)              Temp file (list_file)      Heap file              Aggregate/Analysis

```

1) list_file->tuple_cnt is changed to INT64
- db_query_count(),  db_query_seek_tuple(), insert/delete/update routines in execute_statement.c take list_file->tuple_cnt value. therefore they are changed.
- packing/unpacking at object_representation.c is changed. (aligning is considered)
2) stats that counting # of rows/keys in scan_stats are changed to print correctly for `;trace on`

About client cursor (cursor.c)
- trivial change in cursor.c, typo; when I investigating the code, I found a typo.
- tuple number and tuple count in cursor id don't have to change. They get the number from the page's # of tuples, not the whole list_file's count.


I've tested for the following count queries and showed the trace
```
;sc t1
====
 <Class Name>

     t1

 <Attributes>

     c1                   BIGINT NOT NULL

 <Constraints>

     PRIMARY KEY pk_t1 ON t1 (c1)
```
```
SELECT count(*) from t1;
=== <Result of SELECT Command in Line 3> ===

              count(*)
======================
            2232889346

1 row selected. (1126.773044 sec) Committed.

=== Auto Trace ===

Query Plan:
  TABLE SCAN (t1)

  rewritten query: select count(*) from t1 t1

Trace Statistics:
  SELECT (time: 1126689, fetch: 4885872, ioread: 4885871)
    SCAN (table: t1), (heap time: 896044, fetch: 4885870, ioread: 4885870, readrows: 2232889346, rows: 2232889346)
```
```
> SELECT count(c1) from t1;
=== <Result of SELECT Command in Line 5> ===

             count(c1)
======================
            2232889346

1 row selected. (1800.630586 sec) Committed.

=== Auto Trace ===

Query Plan:
  TABLE SCAN (t1)

  rewritten query: select count(t1.c1) from t1 t1

Trace Statistics:
  SELECT (time: 1800615, fetch: 4885872, ioread: 4885870)
    SCAN (table: t1), (heap time: 1257744, fetch: 4885870, ioread: 4885870, readrows: 2232889346, rows: 2232889346)
```
```
SELECT count(distinct c1) from t1;
=== <Result of SELECT Command in Line 7> ===

    count(distinct c1)
======================
            2232889346

1 row selected. (14114.677945 sec) Committed.

=== Auto Trace ===

Query Plan:
  TABLE SCAN (t1)

  rewritten query: select count(distinct t1.c1) from t1 t1

Trace Statistics:
  SELECT (time: 14114639, fetch: 323099218, ioread: 49867151)
    SCAN (table: t1), (heap time: 1642595, fetch: 4885870, ioread: 4885870, readrows: 2232889346, rows: 2232889346)
```
```
SELECT count(c1) from t1 where c1 > 10000;

=== <Result of SELECT Command in Line 9> ===

             count(c1)
======================
            2232879346

1 row selected. (1606.180724 sec) Committed.

=== Auto Trace ===

Query Plan:
  INDEX SCAN (t1.pk_t1) (key range: (t1.c1> ?:0 ), covered: true)

  rewritten query: select count(t1.c1) from t1 t1 where (t1.c1> ?:0 )

Trace Statistics:
  SELECT (time: 1606117, fetch: 12541203, ioread: 2892376)
    SCAN (index: t1.pk_t1), (btree time: 1144111, fetch: 12541202, ioread: 2892375, readkeys: 2232879346, filteredkeys: 0, rows: 2232879346, covered: true)
```

hgryoo committed:
draft

hgryoo committed:
Merge remote-tracking branch 'upstream/feature/bigint_count' into int64_tuple_count

hgryoo committed:
indeent

hgryoo committed:
revert changing type in cursor, it is one page's tuple count. fix typo

hgryoo committed:
revert, unnecessary

hgryoo committed:
remove redundant comment, add comment for aligning tuple_count

hgryoo committed:
tuple_no indicates tuple position in a page

hgryoo committed:
revert

hgryoo committed:
fix typo

hgryoo committed:
fix indent

beyondykk9 commented on src/query/query_opfunc.c:
```diff
@@ -6307,7 +6307,7 @@ qdata_get_single_tuple_from_list_id (THREAD_ENTRY * thread_p, qfile_list_id * li
   TP_DOMAIN *domain_p;
   char *ptr;
-  int tuple_count, value_count, i;
+  INT64 tuple_count, value_count, i;
```
```suggestion
  INT64 tuple_count;
  int value_count, i;
```

hgryoo replied:
thank you, fixed.

ctshim commented on src/compat/db_query.c:
```diff
@@ -1089,7 +1089,7 @@ db_dump_query_result (DB_QUERY_RESULT * r)
       fprintf (stdout, "Query_id: %lld \n", (long long) r->res.s.query_id);
       fprintf (stdout, "Stmt_id: %d \n", r->res.s.stmt_id);
-      fprintf (stdout, "Tuple Cnt: %d \n", r->res.s.cursor_id.list_id.tuple_cnt);
+      fprintf (stdout, "Tuple Cnt: %lld \n", r->res.s.cursor_id.list_id.tuple_cnt);
```
Is it not necessary to consider the difference between %lld and %I64d?

fprintf (stdout, "Tuple Cnt: %" PRId64 " \n", r->res.s.cursor_id.list_id.tuple_cnt);

hgryoo replied:
I think other codes in CUBRID prefer to print INT64 with "%lld". I'm going to keep it.

hgryoo replied:
I found they cast to (long long int) for INT64, and (unsigned long long int) for UINT64 when printing with `%lld` and `%llu` respectively. I fixed them. Thank you.

beyondykk9 commented on src/compat/db_query.c:
```diff
@@ -2455,8 +2455,8 @@ int
-  int rel1, rel2, rel3, rel_n;
-  int curr_tplno, tpl_cnt;
+  INT64 rel1, rel2, rel3, rel_n;
+  INT64 curr_tplno, tpl_cnt;
```
curr_tplno gets the value from result->res.s.cursor_id.tuple_no.
but the tuple_no is int-type not INT64.
if the tuple_no can be 64 bit value, the type of tuple_no should be changed to INT64 first.

beyondykk9 commented on src/compat/db_query.c:
```diff
@@ -2455,8 +2455,8 @@ int
   int scan;
-  int rel1, rel2, rel3, rel_n;
-  int curr_tplno, tpl_cnt;
+  INT64 rel1, rel2, rel3, rel_n;
```
rel1 is assigned from offset, but the offset is not a INT64 type.
and the offset is passed from fn_execute_internal's max_row which's type is "int" not INT64.
max_row type is from client and the type should be change to INT64 first.

rel2, rel3, rel_n should be also considered.

mhoh3963 commented on src/query/scan_manager.h:
```diff
@@ -304,18 +304,18 @@ typedef struct scan_stats SCAN_STATS;
   struct timeval elapsed_scan;
-  UINT64 num_fetches;
-  UINT64 num_ioreads;
+  INT64 num_fetches;
```
if this values never be minus, how about remain data type as UINT64 ?
If you agree with me, I think that you are better to change the following changed data type to UINT64.

hgryoo replied:
I fixed their data types as UINT64 and printing them by `%llu`.

mhoh3963 reviewed (approved):
LGTM.

hgryoo committed:
casting to long long when printing by lld

hgryoo committed:
set counting scan stats as UINT64 type and printing them by llu

hgryoo committed:
trivial

ctshim commented on src/query/list_file.c:
```diff
@@ -5260,7 +5260,7 @@ qfile_print_list_cache_entry (THREAD_ENTRY * thread_p, FILE * fp, const void *ke
 
   fprintf (fp,
-	   " } tuple_cnt %d page_cnt %d first_vpid { %d %d } last_vpid { %d %d } lasttpl_len %d query_id %lld  "
+	   " } tuple_cnt %lld page_cnt %d first_vpid { %d %d } last_vpid { %d %d } lasttpl_len %d query_id %lld  "
```
How about explicitly type casting to (long long) when outputting to %lld (tuple_cnt %lld)?

hgryoo replied:
I slip it, thank you.

ctshim commented on src/query/query_dump.c:
```diff
@@ -807,7 +807,7 @@ qdump_print_list_id (QFILE_LIST_ID * list_id_p)
     }
 
-  fprintf (foutput, ")(tuple_cnt:%d)", list_id_p->tuple_cnt);
+  fprintf (foutput, ")(tuple_cnt:%lld)", list_id_p->tuple_cnt);
```
How about explicitly type casting to (long long) when outputting to %lld?

hgryoo replied:
I slip it, thank you.

hgryoo committed:
slip, casting for

hgryoo committed:
Modify to have only the INT type for the offset and # of tuples in the db layer.
- Add TODO: higher layers on the db should be extended together to support peeking with offset beyond INT range.

hgryoo committed:
slip, revert to have int type

shparkcubrid commented:
The SQL test results below is needed to check.

sql/_23_apricot_qa/_02_performance/_01_filtered_index/cases/filtered_index_data_type_basic_view.sql - linux_sql_64bit_release
```
-(select count(v_t?_a?c.b), count(v_t?_a?c.c), count(v_t?_a?c.a) from t? v_t?_a?c where (v_t?_a?c.a> ?:? ) and (v_t?_a?c.d> ?:? ) using index v_t?_a?c.i_t?_a?c(+))
+(select count(v_t?_a?c.b), count(v_t?_a?c.c), ( cast(count(v_t?_a?c.a) as integer)) from t? v_t?_a?c where (v_t?_a?c.a> ?:? ) and (v_t?_a?c.d> ?:? ) using index v_t?_a?c.i_t?_a?c(+))
```

hgryoo commented:
> The SQL test results below is needed to check.
> 
> sql/_23_apricot_qa/_02_performance/_01_filtered_index/cases/filtered_index_data_type_basic_view.sql - linux_sql_64bit_release
> 
> ```
> -(select count(v_t?_a?c.b), count(v_t?_a?c.c), count(v_t?_a?c.a) from t? v_t?_a?c where (v_t?_a?c.a> ?:? ) and (v_t?_a?c.d> ?:? ) using index v_t?_a?c.i_t?_a?c(+))
> +(select count(v_t?_a?c.b), count(v_t?_a?c.c), ( cast(count(v_t?_a?c.a) as integer)) from t? v_t?_a?c where (v_t?_a?c.a> ?:? ) and (v_t?_a?c.d> ?:? ) using index v_t?_a?c.i_t?_a?c(+))
> ```

This is what I checked. I'm not well-known about this. Please let me know If the following description is wrong.

By executing the following, a virtual table spec is created with <BIGINT, BIGINT, INT>.
```
create view v_t1_a2c as select /*+ RECOMPILE */ count(b) as num1,count(c) as num2,10 as num3 from t1 where exists (select t1.a) and a>'0' and d > N'0' using index i_t1_a2c(+);
```

Then, When the following queries are added by ADD QUERY Clause, count(a) as num3 is going to be casted according to the virtual table spec that was created at the `create view` previously performed.
```
alter view v_t1_a2c add query select /*+ RECOMPILE */ count(b) as num1,count(c) as num2,count(a) as num3 from t1 where  a>'0'and d > N'0'using index i_t1_a2c(+);
```
